New URL changes

kevintab95 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of NA.
2. This PR does the following:

- Introduces url_fragment field for topics and topic summaries.
- Changes to URLs for new structures.
- Adds validation for topic name (https://github.com/oppia/oppia/issues/9986) and url_fragment field.
- Introduces Classroom domain object.
- Fixes TODO https://github.com/oppia/oppia/issues/10191.


Details:

 | Page | Old Url  | New Url |
|---|---|---|
| Topic viewer page | `/topic/<topic_name>` | `/learn/<classroom_url_fragment>/<topic_url_fragment>` |
| Topic viewer page > Revision (tab) |  `/topic/<topic_name>`  | `/learn/<classroom_url_fragment>/<topic_url_fragment>/revision` |
| Revision page | `/subtopic/<topic_name>/<subtopic_id>` | `/learn/<classroom_url_fragment>/<topic_url_fragment>/revision/<subtopic_url_fragment>`  |
| Topic viewer page > Practice (tab) | `/topic/<topic_name>`  | `/learn/<classroom_url_fragment>/<topic_url_fragment>/practice` |
| Practice session | `/practice_session?selected_subtopic_ids=1,2,3` | `/learn/<classroom_url_fragment>/<topic_url_fragment>/practice/session?selected_subtopic_ids=1,2,3` |
| Topic viewer page > Story (tab) | `/topic/<topic_name>` | `/learn/<classroom_url_fragment>/<topic_url_fragment>/story` |
| Story viewer page | `/story/<story_id>`  | `/learn/<classroom_url_fragment>/<topic_url_fragment>/story/<story_url_fragment>`  |
| Review Test | `/review_test/<story_id>`  | `/learn/<classroom_url_fragment>/<topic_url_fragment>/review-test/<story_url_fragment>`  |


## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

Commit:
M1: new urls

Commit:
fix typescript error

Commit:
fix merge conflicts

seanlip commented on assets/constants.ts:
```diff
@@ -17,6 +17,8 @@ export = {
   // Whether to allow custom event reporting to Google Analytics.
   "CAN_SEND_ANALYTICS_EVENTS": false,
 
+  "CLASSROOM_STATUS_STAGING": "staging",
```
I'm not sure "CLASSROOM_STATUS" is valid for this -- a "status" sounds like something that can be applied to any classroom. Perhaps DUMMY_CLASSROOM_NAME and provide some explanation of what this is in the comments.

kevintab95 replied:
Done.

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -45,6 +47,24 @@
 (suggestion_models,) = models.Registry.import_models([models.NAMES.suggestion])
 
 
+def handle_redirection(handler, redirection_url, expected_return_type):
```
The concept of "permanent" should probably be in both the function name and docstring.

kevintab95 replied:
Done.

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -2305,11 +2325,11 @@ def can_access_topic_viewer_page(handler):
     """
 
-    def test_can_access(self, topic_name, **kwargs):
+    def test_can_access(self, abbrev_topic_name, **kwargs):
```
abbrev --> abbreviated

kevintab95 replied:
Done.

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -2318,11 +2338,47 @@ def test_can_access(self, topic_name, **kwargs):
-        topic_name = python_utils.url_unquote_plus(topic_name)
-        topic = topic_fetchers.get_topic_by_name(topic_name)
+        lowercase_abbrev_topic_name = (
+            python_utils.url_unquote_plus(abbrev_topic_name).replace('-', ' '))
```
Why do you need url_unquote_plus()? There is no indication that abbreviated_topic_name is part of a URL, at least from the variable name.

Also, why don't you just have the "abbreviated name" in the topic store the URL fragment directly, rather than doing all this encoding? It's more direct. You can also have that field include validation to ensure that only a very restricted subset of characters is used.

kevintab95 replied:
Done.

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -45,6 +47,24 @@
 (suggestion_models,) = models.Registry.import_models([models.NAMES.suggestion])
 
 
+def handle_redirection(handler, redirection_url, expected_return_type):
```
Also, try not to start function names with "handle". It is vague. Have them "do" something -- e.g. redirect(), redirect_permanently(), etc.

kevintab95 replied:
Done.

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -2318,11 +2338,47 @@ def test_can_access(self, topic_name, **kwargs):
+            python_utils.url_unquote_plus(abbrev_topic_name).replace('-', ' '))
+        topic = topic_fetchers.get_topic_by_abbreviated_name(
+            lowercase_abbrev_topic_name)
+        classroom_name = kwargs.pop('classroom_name', None)
```
If this is an expected argument, why not pass it in explicitly?

kevintab95 replied:
This is an expected argument for html page requests. For data handler requests we don't need this parameter and that is why I have made it an optional one -- this acl decorator method is being used by the topic data handler too. Ditto for other methods below.

seanlip replied:
Then use two handlers, or have the parameter be required (but unused) even for the data handler. Don't do things implicitly like this.

kevintab95 replied:
Done.

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -2318,11 +2338,47 @@ def test_can_access(self, topic_name, **kwargs):
         if topic is None:
-            raise self.PageNotFoundException
+            handle_redirection(
+                self, '/%s' % classroom_name,
```
What if classroom_name is None?

kevintab95 replied:
classroom_name would be None only for data calls. In which case, the redirect_permanently() method will throw a 404.

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -2318,11 +2338,47 @@ def test_can_access(self, topic_name, **kwargs):
+            return
+        if classroom_name not in [None, constants.CLASSROOM_STATUS_STAGING]:
+            lowercase_classroom_name = (
+                python_utils.url_unquote_plus(
```
Please try and design the fields etc. so as to remove the need for this construction anywhere. It is preferable to avoid string manipulation where possible if it isn't necessary.

kevintab95 replied:
For classroom name, we need this because we do not have a dedicated "url fragment" field for classrooms. So there may be classrooms that are (1) not lowercase, and (2) containing spaces.

seanlip replied:
Then we need to add such a field for classrooms.

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -2318,11 +2338,47 @@ def test_can_access(self, topic_name, **kwargs):
+            lowercase_classroom_name = (
+                python_utils.url_unquote_plus(
+                    classroom_name).replace('-', ' '))
+            topic_ids_from_classroom_pages = (
```
Keep naming consistent; you've changed "for" to "from".

kevintab95 replied:
Done.

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -2318,11 +2338,47 @@ def test_can_access(self, topic_name, **kwargs):
+                        handle_redirection(
+                            self,
+                            '/learn/%s/%s' % (
+                                classroom_dict['name'].lower().replace(
```
Again we need to find a way to not repeat this construct everywhere.

kevintab95 replied:
I have moved this to a separate method above.

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -2318,11 +2338,47 @@ def test_can_access(self, topic_name, **kwargs):
+                        topic_is_assigned_classroom = True
+                        break
+                    else:
+                        handle_redirection(
```
As a note -- these "error" redirections should not be permanent. Do not permanently redirect URLs to an error page. Only have permanent redirections from old URLs to their corresponding new URLs.

kevintab95 replied:
I think it should be permanent if the URL field values passed are invalid. We don't "redirect" to error pages actually. For data requests that contain invalid params we directly throw a 404 (and not redirect).

seanlip replied:
If the URL field values passed are invalid, you redirect "upwards", right? That redirection should not be permanent. As an example, someone may accidentally enter a URL for a chapter that is yet to be written, so the URL may actually become valid someday. You don't want the URL to end up being invalid forever because of a 301 redirect.

Only use a 301 redirect when a page has moved from one explicit place to another explicit place.

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -2367,16 +2426,65 @@ def test_can_access(self, story_id, *args, **kwargs):
+                        story_id),
+                    self.GET_HANDLER_ERROR_RETURN_TYPE)
+                return
+            if classroom_name not in [None, constants.CLASSROOM_STATUS_STAGING]:
```
There seems to be a great deal of repetition in the logic, here and elsewhere. You might want to have a classroom_services.py or similar to handle some of the common logic.

kevintab95 replied:
I have added a common method above. Done.

seanlip commented on core/controllers/acl_decorators_test.py:
```diff
@@ -2144,26 +2148,30 @@ def setUp(self):
+                '/mock_story/topic/%s' % self.story_id,
+                expected_status_int=200)
 
 
```
You should have way more tests than this, given the amount of new code that has been added to acl_decorators.py.

kevintab95 replied:
Done.

seanlip commented on core/controllers/practice_sessions_test.py:
```diff
@@ -46,7 +46,7 @@ def setUp(self):
 
         self.topic = topic_domain.Topic.create_default_topic(
-            self.topic_id, 'public_topic_name', 'abbrev', 'description')
+            self.topic_id, 'public_topic_name', 'public', 'description')
```
To help the reader, add some text to 'public' (or 'private' below) that states what the field represents. Similar to the 'public_topic_name' before it.

Ditto elsewhere.

kevintab95 replied:
Done.

seanlip commented on core/controllers/topic_editor.py:
```diff
@@ -202,13 +204,19 @@ def get(self, topic_id):
             grouped_skill_summary_dicts[topic_object.name] = skill_summary_dicts
 
+        classroom_name = constants.CLASSROOM_STATUS_STAGING
+        for classroom_dict in config_domain.TOPIC_IDS_FOR_CLASSROOM_PAGES.value:
```
Where possible, try to reuse code that is similar in multiple places. Again, you might want to have a classroom_services.py that deals with anything relating to the config property.

kevintab95 replied:
Hm, but isn't this case pretty compact? We are just checking if a value exists in a list of objects. Also, AFAIK we use *_services.py for domain objects. Classroom is not really a domain object -- it is a config property.

seanlip replied:
No, that's not quite right -- some services do not have associated domain objects. In this case, a "classroom" is still a logical concept and should be treated as such. The only reason we don't have a domain object at the moment is because we are "cheating" by using the config property when the right way is really to be using a storage model.

So, please still create classroom_services.py and have anything that is conceptually related to "classroom" be moved there.

kevintab95 replied:
Got it. Done.

seanlip commented on core/domain/topic_domain.py:
```diff
@@ -580,6 +580,30 @@ def require_valid_name(cls, name):
+                'Abbreviated name field should not be empty.')
+        if '-' in name:
+            raise utils.ValidationError(
+                'Abbreviated name field should not contain hyphens.')
```
Perhaps "abbreviated name" is the wrong word, if we mean something like "topic url fragment".

kevintab95 replied:
Removed this. Also, I have renamed all user facing messages to contain "Topic URL fragment" instead of "Abbreviated name".

seanlip commented on core/domain/topic_domain.py:
```diff
@@ -1449,6 +1474,31 @@ def __init__(
+        Args:
+            name: str. The abbreviated name to validate.
+        """
+        if not isinstance(name, python_utils.BASESTRING):
```
Avoid repetition please.

kevintab95 replied:
Done. Moved to utils.

seanlip reviewed (changes requested):
Hi @kevintab95 -- I have taken a partial pass, and my initial impression is that there is far too much code repetition / copy-pasting. Could you please try to consolidate things more? (I left more specific comments inline.)

Thanks!

Showtim3 commented:
Hey @kevintab95 Mind the merge conflicts.

Commit:
Merge remote-tracking branch 'upstream/develop' into new-urls-m1

Commit:
address comments

Commit:
fix lint issues

Commit:
fix lint issues

kevintab95 reviewed (commented):
Addressed comments. PTAL @seanlip!

Commit:
Merge remote-tracking branch 'upstream/develop' into new-urls-m1

Commit:
fix lint issues

U8NWXD reviewed (changes requested):
@kevintab95 the e2e tests look good! I think all that's left is to make sure they aren't flaky. Given how many suites are affected, getting 3 consecutive passes of all the tests might be easier than 5 passes for each suite, but either works for me

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -45,6 +47,72 @@
+    Returns:
+        bool. Whether classroom name provided was valid or not.
+    """
+    lowercase_classroom_name = (
```
Again, don't do implicit conversion. Add a new field in the config property if needed.

kevintab95 replied:
Done.

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -45,6 +47,72 @@
+            else:
+                redirect_permanently(
+                    handler, '/learn/%s/%s' % (
+                        classroom_dict['name'].lower().replace(' ', '-'),
```
Same thing, don't do conversions. Store the data directly and only check equality. I think we will want/need the freedom to define arbitrary url fragments and they may not match the classroom name exactly.

kevintab95 replied:
Done.

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -45,6 +47,72 @@
+                        url_substring),
+                    handler.GET_HANDLER_ERROR_RETURN_TYPE)
+                return False
+    if not topic_is_assigned_classroom:
```
As mentioned before, please don't do permanent redirects for the error case.

kevintab95 replied:
Done.

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -2318,11 +2386,22 @@ def test_can_access(self, topic_name, **kwargs):
 
         if topic is None:
-            raise self.PageNotFoundException
+            redirect_permanently(
```
Ditto as above, please do not redirect any error cases permanently. Here and elsewhere.

kevintab95 replied:
Done.

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -45,6 +47,72 @@
+        handler.redirect(redirection_url, permanent=True)
+
+
+def validate_and_redirect_based_on_classroom_name(
```
This function is doing too many things. Just have it validate; leave the redirection to the decorators. Put the validation function in classroom_services.

kevintab95 replied:
Done.

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -2407,23 +2512,27 @@ def can_access_subtopic_viewer_page(handler):
-        topic = topic_fetchers.get_topic_by_name(topic_name)
+        topic = topic_fetchers.get_topic_by_abbreviated_name(
+            abbreviated_topic_name)
+        classroom_name = kwargs.pop('classroom_name', None)
```
Same comments apply here and below.

kevintab95 replied:
Done.

seanlip commented:
Also added:

https://github.com/oppia/oppia/pull/10041#discussion_r461414299
https://github.com/oppia/oppia/pull/10041#discussion_r461416990
https://github.com/oppia/oppia/pull/10041#discussion_r461415757

PTAL @kevintab95

Commit:
Address comments

kevintab95 reviewed (commented):
Hi @seanlip, this PR is ready for another pass. PTAL!

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -38,13 +39,30 @@
 (suggestion_models,) = models.Registry.import_models([models.NAMES.suggestion])
 
 
+def redirect(handler, redirection_url, expected_return_type):
```
At this point I'm starting to wonder why this function needs to be defined at all. Don't we already have a self.redirect() in each handler?

Or alternatively, if the key is "if not json" then have the function name reflect that.

kevintab95 replied:
Done. Changed it to `redirect_based_on_return_type`.

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -2318,11 +2338,26 @@ def test_can_access(self, topic_name, **kwargs):
+            classroom_services.get_classroom_url_fragment_for_topic_id(
+                topic.id))
+        if classroom_url_fragment != verified_classroom_url_fragment:
+            url_substring = abbreviated_topic_name
```
Probably should rename "abbreviated_topic_name" to "topic_url_fragment" or similar. The meaning seems to have changed. (Or, declare a new field and later deprecate the old one.)

If this involves updating the topic model, that should be fine, we can add a one-off job to populate the new field from the old one, and deprecate the old one later. 

Also you can just use this variable inline below, right? No need to define a separate variable.

EDIT: Thinking about it more -- I'm not even sure whether abbreviated_topic_name and urlsafe_topic_fragment are the same thing. Didn't the Android team need the former for a different purpose (for the summary cards or something like that)? If so, you might need to have both fields as separate things.

kevintab95 replied:
Done. Added a new url_fragment field for topic.

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -2352,10 +2387,15 @@ def can_access_story_viewer_page(handler):
 
-    def test_can_access(self, story_id, *args, **kwargs):
+    def test_can_access(
+            self, story_id, abbreviated_topic_name,
```
Should it follow the order in the url? classroom_url_fragment, topic_url_fragment, story_id?

kevintab95 replied:
Done.

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -2407,23 +2474,29 @@ def can_access_subtopic_viewer_page(handler):
         if the user can access the give subtopic viewer page.
```
give --> given

kevintab95 replied:
Done.

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -2407,23 +2474,29 @@ def can_access_subtopic_viewer_page(handler):
 
         Args:
-            topic_name: str. The name of the topic.
+            abbreviated_topic_name: str. The abbreviated name of the topic
```
Same notes as above re naming.

kevintab95 replied:
Done.

seanlip commented on core/controllers/classroom.py:
```diff
@@ -53,16 +54,13 @@ def get(self, classroom_name):
-                classroom_name_is_valid = True
-                topic_ids = classroom_dict['topic_ids']
-                break
+        classroom_dict = classroom_services.get_classroom_dict_by_name(
```
Can you make a Classroom domain object? In general, shouldn't pass dicts around.

kevintab95 replied:
Created a Classroom domain object. Done.

seanlip commented on core/controllers/topics_and_skills_dashboard.py:
```diff
@@ -251,7 +251,7 @@ class NewTopicHandler(base.BaseHandler):
         """Handles POST requests."""
         name = self.payload.get('name')
-        abbreviated_name = self.payload.get('abbreviated_name')
+        abbreviated_name = self.payload.get('abbreviatedName')
```
This is odd, all these should be snake_case. Please fix the frontend if it's passing the wrong things?

kevintab95 replied:
Done.

seanlip commented on core/domain/classroom_services.py:
```diff
@@ -0,0 +1,57 @@
+        classroom_name: str. The classroom name.
+
+    Returns:
+        dict|None. Returns the classroom dict if found or else returns None.
```
... if found, else ...

kevintab95 replied:
Done.

seanlip commented on core/domain/config_domain.py:
```diff
@@ -70,6 +71,15 @@
                 'type': 'unicode'
             }
+        }, {
+            'name': 'url_fragment',
```
I noticed the structure for this property has changed. Will that cause any issues? (Test by checking out develop, setting something there, switching to this branch, seeing if the existing stuff that uses that property works fine before editing it, then editing the config property, and seeing if the stuff that uses that property works fine after editing it.)

kevintab95 replied:
So I tried this and it looks like the topics and skills dashboard will work as expected with the new config changes. But the url for classroom will not work unless the url_fragment for this classroom in the config property is updated manually. I will send clear instructions in an email once this PR has been merged.

seanlip commented on core/storage/topic/gae_models.py:
```diff
@@ -152,6 +152,22 @@ def get_by_name(cls, topic_name):
+        """
+        return TopicModel.query().filter(
+            cls.abbreviated_name == abbreviated_name.lower()).filter(
+                cls.deleted == False).get() #pylint: disable=singleton-comparison
```
Add space after #

kevintab95 replied:
Done.

seanlip commented on core/templates/pages/topic-editor-page/editor-tab/topic-editor-tab.directive.html:
```diff
@@ -26,13 +26,25 @@
+                 maxlength="<[MAX_CHARS_IN_ABBREV_TOPIC_NAME]>" ng-trim="true">
+          <span class="oppia-input-box-subtitle">
+            <i>
+              Topic URL fragment should be a short word or set of words in lowercase and separated by hyphens (-). Please use meaningful keywords and avoid using words like "and", "of", or "the". This will be used to uniquely access the topic viewer page. It should be at most <[MAX_CHARS_IN_ABBREV_TOPIC_NAME]> characters.
```
The topic URL fragment is used to uniquely access the topic viewer page. It should consist of one or more hyphen-separated words, all in lowercase, with at most `<[MAX_CHARS_IN_ABBREV_TOPIC_NAME]>` characters in total. Please use meaningful keywords, and avoid using words like "and", "of", or "the".

kevintab95 replied:
Done.

seanlip commented on core/templates/pages/topic-editor-page/editor-tab/topic-editor-tab.directive.html:
```diff
@@ -26,13 +26,25 @@
+          <label for="abbreviatedTopicName" class="form-heading protractor-test-abbrev-topic-name-heading">Topic URL Name</label>
+          <input id="abbreviatedTopicName" type="text" class="form-control protractor-test-abbrev-topic-name-field"
+                 ng-disabled="!topicRights.canEditTopic()"
+                 ng-model="editableAbbreviatedName" ng-blur="updateAbbreviatedTopicName(editableAbbreviatedName)"
```
Any validation on allowed chars / regex?

kevintab95 replied:
I've added validation for url fragments in the topic object factory. It will show up in the warnings list. Done.

seanlip commented on core/templates/pages/topics-and-skills-dashboard-page/templates/create-new-topic-modal.template.html:
```diff
@@ -15,6 +15,15 @@ <h3>
+      <input class="protractor-test-new-topic-url-fragment-field" ng-model="newlyCreatedTopic.abbreviatedName" maxlength="<[MAX_CHARS_IN_ABBREV_TOPIC_NAME]>">
+      <div class="oppia-input-box-subtitle">
+        <i>
+          Topic URL fragment should be a short word or set of words in lowercase and separated by hyphens (-). Please use meaningful keywords and avoid using words like "and", "of", or "the". This will be used to uniquely access the topic viewer page. It should be at most <[MAX_CHARS_IN_ABBREV_TOPIC_NAME]> characters.
```
See above notes.

kevintab95 replied:
Done.

seanlip commented on schema_utils_test.py:
```diff
@@ -170,7 +170,12 @@
-        'is_supported_audio_language_code': {}
+        'is_supported_audio_language_code': {},
+        'is_url_fragment': {
+            'char_limit': {
```
Don't we already have an existing validator of the form has_length_at_most()?

kevintab95 replied:
Yep, changed this. Done.

Commit:
Milestone 2: Url Changes
- Adds url_fragment field for story and subtopic.

Commit:
fix typescript errors

Commit:
address comments

Commit:
Merge remote-tracking branch 'upstream/develop' into url-changes-2

Commit:
fix lint issue

Commit:
address comments

Commit:
Merge remote-tracking branch 'upstream/develop' into new-urls-m1

Commit:
address comments (2)

Commit:
fix lint errors

Commit:
fix lint error

Commit:
fix lint error

kevintab95 reviewed (commented):
Note: This PR is not ready for review. Some more changes need to be made once #10134 PR has been merged. Thanks!

U8NWXD commented:
@kevinlee12 FYI you can mark the PR as a draft to indicate it isn't ready for review. The tests should still run, in case you need that

kevintab95 commented:
@U8NWXD done, thanks!

Commit:
address comments

Commit:
Merge remote-tracking branch 'upstream/develop' into url-changes-2

Commit:
fix lint issues

Commit:
increase coverage

Commit:
fix lint issues

Commit:
address comments

Commit:
fix lint issue

Commit:
address doc string comments

Commit:
Merge remote-tracking branch 'upstream/develop' into new-urls-m1

Commit:
Merge branch 'url-changes-2' into new-urls-m1

Commit:
Url changes:
- Utilises story and subtopic url_fragment fields for learner page navigation.
- Adds validation for topic name and topic url_fragment in the UI.
- Fixes #10191.

Commit:
Merge remote-tracking branch 'upstream/develop' into new-urls-m1

Commit:
fix test

Commit:
fix test (2)

Commit:
fix lint issues

Commit:
fix typescript error

kevintab95 commented:
Hi @seanlip, this PR is ready for review pass. PTAL! If you think it is ready for codeowners to take a look, please let me know, I'll assign them too. Thanks!

seanlip commented on assets/constants.ts:
```diff
@@ -17,6 +17,12 @@ export = {
+  // in the URL for topics that are not yet attached to a classroom.
+  "DUMMY_CLASSROOM_URL_FRAGMENT": "staging",
+
+  "DEFAULT_CLASSROOM": "math",
```
Explain what this means (i.e. in what sense is it the default?). Also is it the "classroom name" or the "classroom url fragment"? A classroom is more like a domain object, not a string.

kevintab95 replied:
It is the URL fragment. Done.

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -38,13 +40,31 @@
 (suggestion_models,) = models.Registry.import_models([models.NAMES.suggestion])
 
 
+def redirect_based_on_return_type(
```
Just to check, should this be general behaviour in self.redirect? (Not sure.)

kevintab95 replied:
I do not see a use for this in other files. I have just encapsulated L62-65  in this function because it was repeated in multiple places in this file.

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -72,6 +92,41 @@ def test_can_access(self, *args, **kwargs):
+        if not classroom:
+            redirect_based_on_return_type(
+                self, '/learn/%s' % constants.DEFAULT_CLASSROOM,
+                self.GET_HANDLER_ERROR_RETURN_TYPE)
```
Will statements in the decorated handler that come after this one get executed (and should they be)? (I.e. do you need a return statement after this one?)

kevintab95 replied:
Not it is not expected to execute. I have added a return. Done.

seanlip commented on core/controllers/acl_decorators.py:
```diff
@@ -38,13 +40,31 @@
 (suggestion_models,) = models.Registry.import_models([models.NAMES.suggestion])
 
 
+def redirect_based_on_return_type(
```
Should this be prefixed with an underscore (since it seems to be an helper function internal to this module)?

(Do the same for all such functions -- in general prefer making things private where possible. It helps with managing complexity by not exposing so many functions to the outside.)

kevintab95 replied:
Done.

seanlip commented on core/controllers/admin.py:
```diff
@@ -379,6 +379,8 @@ def _load_dummy_new_structures_data(self):
             question_id_1 = question_services.get_new_question_id()
             question_id_2 = question_services.get_new_question_id()
             question_id_3 = question_services.get_new_question_id()
+            question_id_4 = question_services.get_new_question_id()
```
Just curious -- what's the rationale for adding stuff here, and why are you (below) making them all have the same skill ID? The second change seems to reduce the utility of the generated data a bit, and it conflicts with the docstring. So I kind of suggest leaving things as they are -- unless you have a good reason to change them?

kevintab95 replied:
I had added these to test the practice tab (it shows only if there are 5 questions). But I don't think this is required to be done in this PR. I will revert these changes.

kevintab95 replied:
Done.

seanlip commented on core/controllers/admin_test.py:
```diff
@@ -932,6 +932,7 @@ def test_changing_user_role_from_exploration_editor_to_topic_manager(self):
         topic_id = topic_services.get_new_topic_id()
         self.save_new_topic(
             topic_id, user_id, name='Name',
+            abbreviated_name='name', url_fragment='name',
```
Try to use different args to symbolize different things -- e.g. 'abbrev', 'frag'

kevintab95 replied:
Done.

seanlip commented on core/domain/android_validation_constants.py:
```diff
@@ -36,7 +36,7 @@
 # app.constants.ts as well.
 MAX_CHARS_IN_TOPIC_NAME = 39
-MAX_CHARS_IN_ABBREV_TOPIC_NAME = 12
+MAX_CHARS_IN_ABBREV_TOPIC_NAME = 20
```
Just to check -- in a PR which changes URLs, why is this being changed?

Also I was under the impression we had deprecated the abbreviated topic name field ... why is is still around? Where is it used? Maybe check with @mschanteltc? /cc @aks681

kevintab95 replied:
Initially I hd planned to use abbreviated_name in the URL but later on we decided to use a separate url_fragment field for this. I have reverted this change.

It is not currently used. We have only defined it as a constant in android_validation_constants and in constants (to be in sync).

seanlip commented on core/domain/topic_services.py:
```diff
@@ -187,20 +198,50 @@ def _create_topic(committer_id, topic, commit_message, commit_cmds):
-    Raises:
-        Exception. Topic with same name already exists.
     """
-    existing_topic = topic_fetchers.get_topic_by_name(topic.name)
```
Why delete?

kevintab95 replied:
This check can be done before calling save_new_topic. Like we already do in NewTopicHandler in the topics_and_skills_dashboard controller.

seanlip commented on core/domain/wipeout_service_test.py:
```diff
@@ -496,6 +496,8 @@ def setUp(self):
         self.save_new_topic(
             self.TOPIC_1_ID,
             self.user_1_id,
+            abbreviated_name='topic-one',
```
Here and elsewhere, try to make the abbreviated name and url fragment different in general.

kevintab95 replied:
Done.

seanlip commented on core/storage/topic/gae_models.py:
```diff
@@ -150,7 +153,23 @@ def get_by_name(cls, topic_name):
+            TopicModel|None. The topic model of the topic or None if not
+            found.
+        """
+        return TopicModel.query().filter(
```
Please add a TODO and file an issue to find a way to make this faster. We should perhaps generate the full mapping once and store it in the memcache, or something. But doing a full query every time someone accesses a URL is kind of expensive.

kevintab95 replied:
Done. Created #10210 to track this.

seanlip commented on core/storage/topic/gae_models.py:
```diff
@@ -220,6 +240,8 @@ class TopicSummaryModel(base_models.BaseModel):
     # The description of the topic.
     description = ndb.TextProperty(indexed=False)
+    # The url fragment of the topic.
+    url_fragment = ndb.StringProperty(indexed=True)
```
Do we need a TODO/issue to make this required=True?

kevintab95 replied:
Done.

seanlip commented on core/templates/components/forms/validators/is-url-fragment.filter.spec.ts:
```diff
@@ -0,0 +1,78 @@
+  it('should fail when there are special characters other than hyphen',
+    angular.mock.inject(function($filter) {
+      var filter = $filter(filterName);
+      expect(filter('special%$^chars', args)).toBe(false);
```
It's stronger to have one expectation per special char.

kevintab95 replied:
Done.

seanlip commented on core/templates/components/forms/validators/is-url-fragment.filter.spec.ts:
```diff
@@ -0,0 +1,78 @@
+    angular.mock.inject(function($filter) {
+      var filter = $filter(filterName);
+      expect(filter('url with spaces', args)).toBe(false);
+      expect(filter(' tailing space ', args)).toBe(false);
```
nit: trailing spaces

kevintab95 replied:
Done.

seanlip commented on core/templates/pages/topic-editor-page/modal-templates/create-new-subtopic-modal.template.html:
```diff
@@ -25,17 +25,17 @@ <h3>
-      <em>The subtopic url fragment is not unique.</em>
-    </span>
+    <div class="oppia-input-box-subtitle text-danger" ng-if="$ctrl.editableUrlFragment && $ctrl.subtopicUrlFragmentExists">
+      <i>The subtopic url fragment is not unique.</i>
```
Usually we use "em" for semantic emphasis; i is just for styling. Why did you change it?

Please fix for other instances of the "i" tag.

kevintab95 replied:
In create-new-topic-modal.template and in create-new-topic-modal.template, we use "i" inside "spans" with the class "oppia-input-box-subtitle". I just folllowed the same pattern. Please let me know if this has to be changed to "em" in all of these places too. Thanks!

seanlip replied:
Interesting -- after reading up on "i" vs "em", I'd say use "em" everywhere.

kevintab95 replied:
Done.

seanlip commented on core/templates/pages/topic-editor-page/subtopic-editor/subtopic-editor-tab.component.html:
```diff
@@ -63,9 +63,9 @@ <h3 class="subtopic-card-heading">Review Material</h3>
-              <em>The subtopic url fragment is not unique.</em>
-            </span>
+            <div class="oppia-input-box-subtitle text-danger" ng-if="$ctrl.editableUrlFragment && $ctrl.subtopicUrlFragmentExists">
+              <i>The subtopic url fragment is not unique.</i>
```
Ditto.

kevintab95 replied:
See comment above: https://github.com/oppia/oppia/pull/10041#discussion_r467180195

seanlip reviewed (changes requested):
Thanks, took a pass! Feel free to assign other codeowners.

nithusha21 reviewed (approved):
LGTM for codeowner files!

Commit:
address comments

Commit:
fix test

Commit:
fix test

sagangwee reviewed (approved):
@kevintab95 LGTM for codeowners. Thanks!

vojtechjelinek reviewed (approved):
LGTM!

Commit:
fix backend failures

ankita240796 reviewed (approved):
LGTM from codeowner's perspective, Thanks @kevintab95!

Commit:
Merge remote-tracking branch 'upstream/develop' into new-urls-m1

Commit:
fix fe test

Commit:
fix lint issue

kevintab95 reviewed (commented):
Hi @seanlip, please take another pass. Thanks!

Commit:
Merge remote-tracking branch 'upstream/develop' into new-urls-m1

kevintab95 commented:
@BenHenning @DubeySandeep @aks681 @bansalnitish @srijanreddy98 @Showtim3  PTAL as codeowners. Thanks!

seanlip commented on assets/constants.ts:
```diff
@@ -21,7 +21,9 @@ export = {
-  "DEFAULT_CLASSROOM": "math",
+  // The default classroom URL fragment to use when the provided classroom URL
+  // fragment in the controller is invalid.
+  "DEFAULT_CLASSROOM_NAME": "math",
```
Then call it DEFAULT_CLASSROOM_URL_FRAGMENT

kevintab95 replied:
Done.

seanlip commented on assets/constants.ts:
```diff
@@ -21,7 +21,9 @@ export = {
   // in the URL for topics that are not yet attached to a classroom.
   "DUMMY_CLASSROOM_URL_FRAGMENT": "staging",
```
Maybe CLASSROOM_URL_FRAGMENT_FOR_UNATTACHED_TOPICS or similar.

(sorry, bit long, but trying to distinguish from DEFAULT_CLASSROOM_URL_FRAGMENT)

kevintab95 replied:
Done

Commit:
address comments

Commit:
address comments

kevintab95 reviewed (commented):
PTAL @seanlip!

Commit:
fix e2e test for classroom page suite

Commit:
fix classroomPageFileUploadFeatures e2e suite failure

BenHenning commented on core/domain/android_validation_constants_test.py:
```diff
@@ -28,6 +28,9 @@ class AndroidValidationConstantsTest(test_utils.GenericTestBase):
     """Tests verifying the character limits."""
 
     def test_that_character_limits_in_both_files_are_equal(self):
+        self.assertEqual(
```
Thanks for adding the additional test!

BenHenning reviewed (approved):
Approval for Android constant test change (codeowners-only approval).

srijanreddy98 reviewed (approved):
LGTM for code owner files!

Commit:
fix backend failures

Commit:
Merge remote-tracking branch 'upstream/develop' into new-urls-m1

Commit:
change to async ...Async for then-able functions in ClassroomBackendApiService

kevintab95 commented:
The frontend coverage check for this PR seems to fail on github actions but passes locally -- karma coverage report shows 100% statement coverage for the reported file. Also, this PR does not make changes to the file (`player-position.service.ts`) mentioned in the report.

bansalnitish reviewed (approved):
LGTM!

seanlip commented:
Hi @kevintab95 -- see #10215, is it relevant? (For the frontend tests question.)

kevintab95 commented:
Yup, #10215 should fix it. Thanks!

Commit:
fix backend coverage issue

Commit:
Merge remote-tracking branch 'upstream/develop' into new-urls-m1

kevintab95 commented:
Hi @U8NWXD, PTAL 3 consecutive passes as requested (https://github.com/oppia/oppia/pull/10041#pullrequestreview-456234910):
![3consecpasses](https://user-images.githubusercontent.com/11008603/89704520-f9945080-d971-11ea-80b8-12d74299b2d5.png)

Showtim3 commented on core/templates/pages/topic-editor-page/services/topic-editor-state.service.ts:
```diff
@@ -429,6 +455,46 @@ angular.module('oppia').factory('TopicEditorStateService', [
+        return _classroomUrlFragment;
+      },
+
+      changeTopicWithNameExists: function(topicName, successCallback) {
```
Hey @kevintab95 Should this be ```checkTopicWithNameExists``` ?

kevintab95 replied:
Hm, it actually updates the variable "topicWithNameExists" so change seemed more appropriate.

seanlip replied:
But semantically the wording doesn't make sense, and ditto for setTopicWithNameExists, etc. -- that doesn't really make sense as a clause in English, so it leads to confusion. I think what you really want to say is "change the _topicWithUrlFragmentExists property" but it reads as "change the topic, under certain conditions". So @Showtim3 has a point here.

It's hard to find a good name. How about doesTopicWithNameExistAsync()? The fact that it updates an inner "caching" variable behind the scenes seems fine, other functions do this too and don't call it out. I would also just drop the setter functions for these "inner" variables and inline that functionality.

One other note is that you should probably design this so that it updates a variable which is used directly (or via a getter) as the source of truth for the template, or return a value that can be used as the source of truth, but not both. The former is probably better, in which case you could call this function checkExistenceOfTopicName() and not have it have a callback. Do you think that would be possible?

kevintab95 replied:
Notes from discussion with @seanlip :

With the current implementation "changeTopicWithNameExists" does not _return_ anything. It does 2 things:

1. Sets _topicWithNameExists. It does so by making a backend call that returns a boolean value. This boolen value is used to set _topicWithNameExists.
2. Executes the success callback. It does not pass any value to the success callback -- the callback being executed just indicates that _topicWithNameExists has been updated.

In order to get the latest value of _topicWithNameExists, the corresponding getter method "getTopicWithNameExists" should be used. Therefore, getTopicWithNameExists is always the source of truth.

That said, I will be changing the names used to avoid confusion and add jsdoc for the same.

Proposed name:

`changeTopicWithNameExists` -> `updateExistenceOfTopicName`

Commit:
increase backend coverage

Commit:
docstrings and method name changes

DubeySandeep reviewed (approved):
LGTM for the code owner files!

DubeySandeep reviewed (approved):
LGTM for the codeowner files!

aks681 reviewed (approved):
Lgtm as codeowner

U8NWXD reviewed (approved):
Thanks for the screenshots @kevintab95. This looks great!
